### PR TITLE
Fix Issue 7016 - local import does not create -deps dependency

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/ddmd/mars.d
Comment (andralex) [inline src/ddmd/mars.d]: space before `(`

File: src/ddmd/mars.d
Comment (andralex) [inline src/ddmd/mars.d]: I think we seldom do vertical alignment, and only for long chains if ever

Comment (andralex) [inline src/ddmd/mars.d]: make it immutable

**Review by andralex — COMMENTED** [review]
So this is the code that was eating too much memory for phobos? If so, it seems it'll be difficult to use for large projects.

**Comment by andralex** [comment]
Thinking of it, it seems the right approach so it's worth enabling, to improve the memory situation globally later on.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Should be `StringTable* importArray`, since in D the `*` is left-associative.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Same here.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: DMD style does not put a space after a cast.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Weird spacing.

File: test/compilable/extra-files/depsOutput9948a.d
Comment (yebblies) [inline test/compilable/extra-files/depsOutput9948a.d]: Unrelated change, should not be in this PR.

File: test/compilable/extra-files/rdepsOutput.sh
Comment (yebblies) [inline test/compilable/extra-files/rdepsOutput.sh]: Does this actually require a shell test?  If possible, it should make use of the TEST_OUTPUT sections which automatically match the compiler output.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: This function does not look like it belongs in `mars.d`, which is not common between the three frontends.  Most likely it belongs in `dmodule.d`.

**Review by yebblies — CHANGES_REQUESTED** [review]

File: src/ddmd/mars.d
Comment (wilzbach) [inline src/ddmd/mars.d]: ... but Dstyle, which is used for Phobos, does: http://dlang.org/dstyle.html

File: src/ddmd/mars.d
Comment (wilzbach) [inline src/ddmd/mars.d]: Required by Dstyle, which is used for Phobos: http://dlang.org/dstyle.html

File: src/ddmd/mars.d
Comment (wilzbach) [inline src/ddmd/mars.d]: At same point we should enable Dscanner for DMD too...

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: I'm aware, but this isn't phobos and the rules are different.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: ??
I'm referring to the spacing around operators: `for (int i = 1; i < m.aimports.dim; i++)`

Comment (wilzbach) [inline src/ddmd/mars.d]: Ah sorry - I read to quickly and thought you meant the space after `for` - `for (`

Comment (yebblies) [inline src/ddmd/mars.d]: Actually, another issue is that this is using `int` for the index when it should be `size_t`.

File: src/ddmd/mars.d
Comment (wilzbach) [inline src/ddmd/mars.d]: Hmm ok, wouldn't it make sense to unify the rules then?
AFAIK most of the rules at DMD are based on the previous C++ style?
CC @andralex

Comment (yebblies) [inline src/ddmd/mars.d]: If we had a time machine, yes.  Re-formatting dmd's source at this point would not add value.

The DMD rules are based off what the 'prevailing' D style was at the time.  Since the style guide didn't specify for space after `cast` I counted the number of each type in phobos, which was fairly evenly split.  So I picked the one that closest matched the existing C++ style.

Since then Andrei decided/declared that a space should be added in phobos code, but IMO that has very little bearing on what style should be used in DMD.

File: src/ddmd/mars.d
Comment (wilzbach) [inline src/ddmd/mars.d]: ... which could have been prevented if the D-idiomatic `foreach (i; 1 .. m.airports.dim)` would have been used.

File: src/ddmd/mars.d
Comment (andralex) [inline src/ddmd/mars.d]: Yah, the styles are quite different so that would be a large project. I think it would be a palpable possibility once we have a complete pretty printer with LaTeX's algorithm for line breaking.

File: src/ddmd/mars.d
Comment (RazvanN7) [inline src/ddmd/mars.d]: At the moment it is used only in this module, so I don't see why it should be moved. If someone finds it to be useful for other purposes it can be moved later on, otherwise it should stick to the place where it is used. I think that the best approach would be to define a common.d file which implements methods that are for general purpose use, but this would mean to go through the whole dmd front end and cherry pick those methods.

File: src/ddmd/mars.d
Comment (RazvanN7) [inline src/ddmd/mars.d]: It's probably best to stick with the dmd style for now.

File: test/compilable/extra-files/rdepsOutput.sh
Comment (RazvanN7) [inline test/compilable/extra-files/rdepsOutput.sh]: I took inspiration from the single deps test that I could find in the test suite

File: test/compilable/extra-files/depsOutput9948a.d
Comment (RazvanN7) [inline test/compilable/extra-files/depsOutput9948a.d]: I don't think that this should be the subject a PR.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: This is incorrect.  What you are missing is that there are two other compiles (LDC and GDC) and that `mars.d` is not common to all three frontends.

**Comment by RazvanN7** [comment]
@yebblies @andralex thank you for your reviews. Hope this is good shape now.

File: test/compilable/extra-files/depsOutput9948a.d
Comment (yebblies) [inline test/compilable/extra-files/depsOutput9948a.d]: Then don't make a PR for it.  Either way it should not be in _this_ PR.

In general, enforcing style in the test suite is actually counterproductive - it reduces coverage of code styles that must be supported by the compiler, even if they are discouraged.

File: src/ddmd/mars.d
Comment (RazvanN7) [inline src/ddmd/mars.d]: Then the functionality of deps is missing from those, in which case the function is useless for those projects

File: test/compilable/extra-files/rdepsOutput.sh
Comment (yebblies) [inline test/compilable/extra-files/rdepsOutput.sh]: OK

File: test/compilable/extra-files/depsOutput9948a.d
Comment (RazvanN7) [inline test/compilable/extra-files/depsOutput9948a.d]: Ok, I dropped the change

Comment (yebblies) [inline test/compilable/extra-files/depsOutput9948a.d]: Thankyou.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: No, they hand-maintain a copy of much of the code in `mars.d`.  Putting this function in `mars.d` when it contains only backend-independent functionality will force them to do the same with this code.  Besides, this function logically belongs with the other module code, doesn't it?  It's not like putting it here allows mars not to depend on module.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Using a `StringTable` for this is rather clunky - it is designed for string interning which isn't actually necessary here - the `Module` instances themselves should already be unique.  The usual way to do this is to instead use `AA *` (`root.aav`) which uses the `Module` instance address as the hash.

Alternatively (and preferably) if this only needs to be run on `Modules` that haven't had semantic3 run, there is a member of `Module` that keeps track of which semantic has been run (`semanticRun`) which can be inspected to see if it has already been covered.  This will be much more efficient.

**Review by yebblies — CHANGES_REQUESTED** [review]

File: src/ddmd/mars.d
Comment (RazvanN7) [inline src/ddmd/mars.d]: Nice! Thanks. I moved the function to dmodule.d and replaced the AA with a check to semanticRun.

Comment (RazvanN7) [inline src/ddmd/mars.d]: Any other comments or ideas?

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Uh...

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Umm...

File: src/ddmd/dmodule.d
Comment (yebblies) [inline src/ddmd/dmodule.d]: Great!   This looks much better.

**Review by yebblies — COMMENTED** [review]
I'm happy with the implementation now, modulo those two unrelated changes that are left.

@CyberShadow @MartinNowak Can you confirm this sufficiently fixes the problem and we're ok with the performance regression from the extra semantic?

**Comment by CyberShadow** [comment]
Currently this fails to build (see CI).

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Still no.  Stop putting unrelated changes in your PRs.

File: src/ddmd/mars.d
Comment (yebblies) [inline src/ddmd/mars.d]: Remove.

**Review by yebblies — CHANGES_REQUESTED** [review]

**Comment by RazvanN7** [comment]
Thanks @yebblies for this extensive review. Hope the PR is in shape for approval.

**Review by yebblies — APPROVED** [review]

**Comment by yebblies** [comment]
Yeah everything looks right with this now. Just needs commits squashed and approval for the behavior change.

**Comment by andralex** [comment]
Cool beans. Thanks to those involved in getting this done in proportion to their respective contributions.

**Review by andralex — APPROVED** [review]

**Comment by CyberShadow** [comment]
> @CyberShadow @MartinNowak Can you confirm this sufficiently fixes the problem and we're ok with the performance regression from the extra semantic?

I haven't done this yet, sorry. Will try to today or tomorrow.

File: src/ddmd/dmodule.d
Comment (WalterBright) [inline src/ddmd/dmodule.d]: It would be nice to have a comment saying what this function does and a rationale for it.

Comment (WalterBright) [inline src/ddmd/dmodule.d]: It also should be `final`.

Comment (yebblies) [inline src/ddmd/dmodule.d]: It's a free function, it's already final.

**Comment by wilzbach** [comment]
> @CyberShadow @MartinNowak Can you confirm this sufficiently fixes the problem and we're ok with the performance regression from the extra semantic?

FWIW ideally we would have a CI check that warns us automatically if there's a regression in performance or build size like CodeCov does currently. @CyberShadow's AreWeSlim was pretty close to it. Anyways this should be a separate discussion and I will initiate one tonight.

File: src/ddmd/dmodule.d
Comment (CyberShadow) [inline src/ddmd/dmodule.d]: Coverage shows that this `return` is never executed. Is the `if` needed?

File: src/ddmd/dmodule.d
Comment (CyberShadow) [inline src/ddmd/dmodule.d]: Ditto

**Comment by CyberShadow** [comment]
> @CyberShadow @MartinNowak Can you confirm this sufficiently fixes the problem

This does fix issue 7016 as stated, however it will still not fix rdmd. rdmd does not use `-deps` output - it uses `-v` (I think because `-v` includes more information such as which libraries are mixed in, which compiler configuration file is used, and which files are read using `mixin` expressions). @RazvanN7 Do you know what would it take to also fix `-v` output?

> and we're ok with the performance regression from the extra semantic?

The difference in performance seems to be negligible across a few tests (times are best of 5):

| master |  #6748 | task                                           |
|--------|--------|------------------------------------------------|
|  6.537 |  6.554 | time rdmd --force --build-only -allinst digger |
|  1.294 |  1.294 | time dmd -o- digger.d                          |
| 15.366 | 15.480 | time make -f posix.mak DMD=dmd # phobos        |

File: src/ddmd/dmodule.d
Comment (RazvanN7) [inline src/ddmd/dmodule.d]: That line is surely covered by the test, I tested locally. I don't know why codecov says it isn't.

Comment (CyberShadow) [inline src/ddmd/dmodule.d]: Do you see it covered in the `-cov` output?

**Comment by RazvanN7** [comment]
@CyberShadow As I see it, -v doesn't need any fixing; rdmd needs to be adapted so that it uses -deps.

**Comment by CyberShadow** [comment]
> rdmd needs to be adapted so that it uses -deps

That may be true; however, I'm not sure that doing so would be less work than changing `-v` output to match. What do you think?

File: src/ddmd/dmodule.d
Comment (RazvanN7) [inline src/ddmd/dmodule.d]: I manually ran the test locally and put some prints to make sure

Comment (CyberShadow) [inline src/ddmd/dmodule.d]: Okay, but I think we still should figure out why CodeCov says it's uncovered. Knowing whether it's covered in the `-cov` output would tell us whether it's a bug in the test suite or in the CodeCov integration.

**Comment by RazvanN7** [comment]
@CyberShadow I haven't seen how the rdmd code looks so at the moment I can't say for sure which one is easier to modify

**Comment by CyberShadow** [comment]
> @CyberShadow I haven't seen how the rdmd code looks so at the moment I can't say for sure which one is easier to modify

Currently rdmd relies completely on `-v` output for everything. Using `-deps` would almost surely mean that rdmd would need to use both `-v` and `-deps` (for reasons mentioned above).

How much work would fixing `-v` involve?

**Comment by RazvanN7** [comment]
@CyberShadow rdmd currently correctly outputs the dependencies and from what you are saying is uses the -v switch from the compiler. Since this fix doesn't affect rdmd and rdmd correctly outputs deps, what is the problem that needs fixing?

**Comment by ibuclaw** [comment]
In my opinion, -deps should be removed.  It was added to support a dead D1 build tool, and I am not aware of any other user.

**Comment by RazvanN7** [comment]
@ibuclaw at DConf there were a few folks who approached me to fix this stating that this switch is essential to them.

**Comment by ibuclaw** [comment]
Who? The rdmd tool wasn't a user when I last checked.

**Comment by CyberShadow** [comment]
rdmd doesn't use `-deps`, it uses `-v`.

> @CyberShadow rdmd currently correctly outputs the dependencies and from what you are saying is uses the -v switch from the compiler. 

I don't understand what you're saying. How can it output the dependencies correctly if the dependency information it gets from dmd's `-v` output is incomplete?

To clarify - the ultimate goal is to take the test case from issue 7016 (`a.d`, `b.d` and `c.d`) and make it so that `rdmd a.d` works. Currently it fails to link because it's not linking in `c`.

**Comment by RazvanN7** [comment]
When I worked on the issue a few months ago, people in the community said that rdmd -deps works as expected, but  I didn't tested.

**Comment by wilzbach** [comment]
>> at DConf there were a few folks who approached me to fix this stating that this switch is essential to them.
> When I worked on the issue a few months ago, people in the community said that rdmd -deps works as expected, but I didn't tested.

FWIW the best way to find them is a small post on the NG ;-)

**Comment by CyberShadow** [comment]
> When I worked on the issue a few months ago, people in the community said that rdmd -deps works as expected, but I didn't tested.

That can't possibly be true as stated above, and as you can check for yourself. They may have been talking about a different issue.

**Comment by RazvanN7** [comment]
@CyberShadow OK. Either way, there is nothing to fix at -v. As I see it, rdmd uses -v and parses the output to gather the information it needs. The -v switch in the compiler is implemented as a set of if branches which print information if -v is used. In my opinion, the cleanest and least time consuming solution is to modify rdmd., otherwise there is nothing to fix at -v

**Comment by CyberShadow** [comment]
> The -v switch in the compiler is implemented as a set of if branches which print information if -v is used.

Right...

> Either way, there is nothing to fix at -v. 

I don't think we're on the same page here. 

Currently `-v` output with the 7016 test case does not mention `c` at all. The question is how hard would it be to get it to print a line like `import    c	(c.d)` in addition to the information about the `b` import. Then rdmd would be fixed as well.

**Comment by RazvanN7** [comment]
@CyberShadow Oh I see. Then a print could be added when semantic3 is performed on dependencies. That will probably do it

**Comment by rainers** [comment]
>  It was added to support a dead D1 build tool, and I am not aware of any other user.

Visual D uses it to track dependencies, but the generated files have become so bloated that analyzing them sometimes takes a considerable part of the build time.

The new VC project integration uses generic file system tracking instead. I was already considering switching to this for the usual projects, too.

BTW: I don't like interpreting -v output as it can easily be fooled by pragma(msg).

**Comment by CyberShadow** [comment]
> BTW: I don't like interpreting -v output as it can easily be fooled by pragma(msg).

Not an issue (at least with dmd) because `-v` output goes to stdout, and `pragma(msg)` goes to stderr.

**Comment by rainers** [comment]
> Not an issue (at least with dmd) because -v output goes to stdout, and pragma(msg) goes to stderr.

I didn't know that, thanks. Just grepped the source and there seem to be a couple of "out of memory" errors that go to stdout instead of stderr. These should be fixed to not hide them when redirecting stdout only.

**Comment by RazvanN7** [comment]
@CyberShadow As I stated above, I think it would be best to change rdmd. The fact is that in order to get the dependencies, semantic3 is called recursively on imported modules. When the -v switch is used, semantic3 is called only on the module which is passed as argument to dmd. This means that if I were to modify -v I would have to call semantic3 on dependencies even if -deps isn't used, in order to print the information needed for rdmd. That's why I think it would make more sense to modify rdmd to use the -deps switch in the compiler.

File: src/ddmd/dmodule.d
Comment (RazvanN7) [inline src/ddmd/dmodule.d]: Does the -cov switch work with tests that are not under the form of unittests?

**Review by andralex — APPROVED** [review]
OK it looks like this is good to go. Still investigating the coverage issues.

**Comment by MartinNowak** [comment]
Well, this has some serious performance implications, but at least it's only active with `-deps`.
Also something to remember that semantic with `-deps` is similar to `-inline` when debugging future linker errors.

**Comment by rainers** [comment]
I understand why this is necessary for rdmd (finding dependencies transitively), but for a build tool (e.g. Visual D) that only wants to detect whether an object file needs to be recompiled (with an unchanged command line), -deps now emits more dependencies than necessary.

**Comment by andralex** [comment]
@rainers now we have: -v which shows exactly the dependencies actually opened while compiling code, and -deps which exhaustively lists dependencies. Is one of these what you need?

**Comment by rainers** [comment]
@andralex Last time time I looked at it, -v was even more verbose than -deps, and the latter already can be expensive to analyze (e.g. with single file compilation the dependency files are often larger than the object files). I currently plan to use generic file access tracking tools instead (this is already done for linking).

**Comment by CyberShadow** [comment]
> I understand why this is necessary for rdmd (finding dependencies transitively),

rdmd doesn't use `-deps`, it uses `-v`. Because of this, this PR does not fix rdmd.

> but for a build tool (e.g. Visual D) that only wants to detect whether an object file needs to be recompiled (with an unchanged command line), -deps now emits more dependencies than necessary.

Visual D probably also needs to know when an import was added/removed inside a function.

> now we have: -v which shows exactly the dependencies actually opened while compiling code,

No, `-v` still doesn't show imports in functions. `-v` remains "broken".

**Comment by CyberShadow** [comment]
> The difference in performance seems to be negligible across a few tests (times are best of 5):

I didn't realize the new functionality was only activated when `-deps` is specified.

**Comment by CyberShadow** [comment]
> I didn't realize the new functionality was only activated when -deps is specified.

Well, looking at the diff, I don't see where that can be true.

@MartinNowak What performance implications?

**Comment by ibuclaw** [comment]
> Well, looking at the diff, I don't see where that can be true.
```
     if (global.params.moduleDeps)
```

**Comment by ibuclaw** [comment]
Why not instead fix `-v` and deprecate `-deps`?

**Comment by ibuclaw** [comment]
> Last time time I looked at it, -v was even more verbose than -deps, and the latter already can be expensive to analyze

I found it to be the opposite.  `-deps` emits all conditional debug and version identifiers, all imports - repeatedly, not just once on module load, and writes out string imports twice.

**Comment by CyberShadow** [comment]
>     if (global.params.moduleDeps)

OK, so it's not a new thing.

I'll redo the tests <_<

**Comment by rainers** [comment]
> Visual D probably also needs to know when an import was added/removed inside a function.

If the import was added, the file already read by the compiler has changed, so recompilation is necessary. If the import itself is not read by the compiler (because the function body is never analyzed) then there is no dependency to it for the currently compiled object file and recompilation is unnecessary. The respective code is compiled elsewhere, i.e. in another object file or library.

> I found it to be the opposite. -deps emits all conditional debug and version identifiers, all imports - repeatedly, not just once on module load, and writes out string imports twice.

Yeah, the duplicates are horrible, but I never saw debug or version identifiers in there. I thought that the same would be printed with -v, but imports seem to be unique. Instead other unrelated stuff is printed like each function, inlining, semantic phases, etc.

**Comment by CyberShadow** [comment]
> I'll redo the tests 

| master |  #6748 | task (`time dmd -o- -deps=/dev/null ...`)                                          |
|--------|--------|------------------------------------------------|
|  1.27 |  3.83 | Digger (`digger.d`) |
|  0.28 | 1.76 | RABCDAsm (`rabcdasm.d`) |
|  1.41 | 3.04 | Phobos (`unittest.d`) |

Yeah, it's pretty bad.

I have to wonder whether the inconvenience of having to use static / top-level imports to avoid https://issues.dlang.org/show_bug.cgi?id=7016 outweighs having to deal with tripling of compilation times.

Should we revert this?

**Comment by andralex** [comment]
We have a suboptimal solution to a problem - let's improve the solution.

**Comment by CyberShadow** [comment]
Is there a plan for improving it? Is it something that *can* be improved, considering the problem?

**Comment by RazvanN7** [comment]
Yes. Parsing is very cheap (I just parsed the entire phobos in 0.5 seconds). What we can do, after the parser library is done, is implement an import visitor which prints imports at parsing time. There is no need to do semantic in order to identify imports.

**Comment by RazvanN7** [comment]
I'm implementing the earlier mentioned visitor as we speak just to test my transitive visitor.

**Comment by CyberShadow** [comment]
That sounds pretty hacky, and I guess won't work with the `from` idiom (which IIRC was what brought attention to this bug in the first place).

**Comment by mathias-lang-sociomantic** [comment]
> There is no need to do semantic in order to identify imports.

Except it won't work with string mixin, mixin templates, `version` & co

**Comment by andralex** [comment]
>Is there a plan for improving it? Is it something that can be improved, considering the problem?

If we revert this for sure there wouldn't be.

**Comment by CyberShadow** [comment]
Well, some ideas are:

- Make this behaviour optional, e.g. with `-vrecursive-deps`
- Move forward with integrating `rdmd` into `dmd`, which would solve one problem affected by issue 7016
- Look into similar alternative solutions for other current users of `-deps` affected by 7016

**Comment by CyberShadow** [comment]
> If we revert this for sure there wouldn't be.

The biggest concern is that unless this is reverted before the next release, we will be stuck with the slowness (as then reverting would be a regression). Reverting it now will give us more options and time to consider possible solutions.

Furthermore, I am not aware of anything blocked by this bug (especially since it does not fix rdmd), but the increase in compilation times will apparently be felt by many users.

**Comment by andralex** [comment]
My understanding is:

* Only clients that use -deps would be affected by longer build times
* The same clients that would be affected by longer build times are currently being affected by incorrect results
* Computing the correct result for -deps entails more work

Is that correct?

**Comment by CyberShadow** [comment]
> Only clients that use -deps would be affected by longer build times

My understanding as well

> The same clients that would be affected by longer build times are currently being affected by incorrect results

Apparently not true, see [rainers' comment above](#issuecomment-302790728)

> Computing the correct result for -deps entails more work

Seems like it, but the amount of work is considerably larger (2×-6× more time).

**Comment by rainers** [comment]
> Computing the correct result for -deps

Maybe it has to be clarified what "correct" means. The documentation is not very explicit, all it says is "print module dependencies (imports/file/version/debug/lib)".
Regarding imports, the old behaviour collects dependencies to **compile** the module, the new behaviour collects imported modules needed to **link** the module.

BTW: I have removed usage of -deps in Visual D for the next release.

**Comment by ibuclaw** [comment]
> BTW: I have removed usage of -deps in Visual D for the next release.

Excellent.

File: src/ddmd/dmodule.d
Comment (MartinNowak) [inline src/ddmd/dmodule.d]: No, we shouldn't have null modules anywhere ;).

**Comment by MartinNowak** [comment]
And not surprisingly bluntly running semantic3 introduced a regression.
[Issue 17601 – [REG2.075.0-b1] segmentation fault for dmd -deps](https://issues.dlang.org/show_bug.cgi?id=17601)
Time to reevalutate this change? It's a long-standing issue that none of the build tools does an optimal rebuild w/ dependency updates. But worse the different approaches hardly improve the root cause.

This PR being no exemption, it effectively only compiles the object files given on the command line, but analyses __all__ imported modules, thus produces recursive dependencies without recursive recompilation. Lots of extra effort, little outcome.
As mentioned by Rainer (https://github.com/dlang/dmd/pull/6748#issuecomment-302790728), shallow dependencies are more useful for incremental recompilation. Simply use the previous dependency information to recompile changed modules and dependents, while simultaneously updating their shallow dependencies.

Personally I'm in favour of reverting this and addressing the topic in the "faster incremental rebuilds" story (https://wiki.dlang.org/Vision/2017H2#Technical).

File: src/ddmd/dmodule.d
Comment (andralex) [inline src/ddmd/dmodule.d]: I think that happens if the code originates as a string (we don't have a published API for that, but we will).

**Comment by ibuclaw** [comment]
I wouldn't be against a revert either.

**Comment by joakim-noah** [comment]
I haven't read this entire comment thread, but since people were still debating this pull up till a couple months ago, I'll note that I was just able to fix [this recently reported regression in dmd 2.075](https://issues.dlang.org/show_bug.cgi?id=17898) by reverting this pull, which I found by bisecting dmd with the test case given. I don't know how realistic it is to want the deps for a unit test build, but at least one user is doing it, and this pull causes memory usage to balloon up and eventually crash dmd on Arch linux/x64 for me in that situation.

**Comment by WalterBright** [comment]
@RazvanN7  This caused a regression: https://issues.dlang.org/show_bug.cgi?id=17898

**Comment by MartinNowak** [comment]
Well this PR also threw away a working mechanism for shallow deps (as needed for make, cmake, ninja et.al.).

**Pull request merged**
